akka-http/src/main/scala/akka/http/scaladsl/server/directives/RouteDirectives.scala

```diff
@@ -82,6 +83,47 @@ trait RouteDirectives {
   def handle(handler: HttpRequest => Future[HttpResponse]): StandardRoute =
     { ctx => handler(ctx.request).fast.map(RouteResult.Complete)(ctx.executionContext) }
 
+  /**
+   * Handle the request using a function.
+   *
+   * @group route
+   */
+  def handleSync(handler: HttpRequest => HttpResponse): StandardRoute =
+    { ctx => FastFuture.successful(RouteResult.Complete(handler(ctx.request))) }
+
+  /**
+   * Handle the request using an asynchronous partial function.
+   *
+   * This directive can be used to include external components request processing components defined as PartialFunction
+   * (like those provided by akka-grpc) into a routing tree defined as routes.
+   *
+   * @param rejections The list of rejections to reject with if the handler is not defined for a request. By default a request
+   *                   is rejected with an empty list of rejections in that case, which is interpreted as "Not Found".
+   *
+   * @group route
+   */
+  def handlePF(handler: PartialFunction[HttpRequest, Future[HttpResponse]], rejections: Seq[Rejection] = Nil): StandardRoute =
+    { ctx =>
+      if (handler.isDefinedAt(ctx.request)) handler(ctx.request).fast.map(RouteResult.Complete)(ctx.executionContext)
```

Comment:
unfortunately `isDefinedAt` is evil, because in some implementations it will actually evaluate the function to determine if it is defined for a certain input (so in this case it'd be evaluated twice: once by `isDefinedAt` and once by the `apply`). `applyOrElse` should work though?

Reply:
Thanks. I changed that. The problem is a bit that you cannot cheaply apply a post-processing step in `applyOrElse`, so you will have to use `andThen` before the `applyOrElse` which isn't quite ideal (but might still be better than running the filter twice).